src/switchroot/ostree-prepare-root.c

```diff
@@ -569,18 +598,58 @@ main (int argc, char *argv[])
         }
     }
 
+  g_autofree char *etc_config = NULL;
+  if (!ot_keyfile_get_value_with_default (config, SYSROOT_KEY, ETC_KEY, "persistent", &etc_config,
+                                          error))
+    errx (EXIT_FAILURE, "failed to parse %s.%s: %s", SYSROOT_KEY, ETC_KEY, error->message);
+  bool etc_transient = FALSE;
+  if (g_str_equal (etc_config, "persistent"))
+    etc_transient = false;
+  else if (g_str_equal (etc_config, "transient"))
+    etc_transient = true;
+  else
+    errx (EXIT_FAILURE, "Invalid %s.%s: %s", SYSROOT_KEY, ETC_KEY, etc_config);
+
+  // In theory these could be distinct, but no reason to try to support it.
+  if (etc_transient && !sysroot_readonly)
+    errx (EXIT_FAILURE, "Must specify %s.%s for %s.%s=transient", SYSROOT_KEY, READONLY_KEY,
+          SYSROOT_KEY, ETC_KEY);
+
   /* Prepare /etc.
    * No action required if sysroot is writable. Otherwise, a bind-mount for
    * the deployment needs to be created and remounted as read/write. */
   if (sysroot_readonly || using_composefs)
     {
-      /* Bind-mount /etc (at deploy path), and remount as writable. */
-      if (mount ("etc", TMP_SYSROOT "/etc", NULL, MS_BIND | MS_SILENT, NULL) < 0)
-        err (EXIT_FAILURE, "failed to prepare /etc bind-mount at /sysroot.tmp/etc");
-      if (mount (TMP_SYSROOT "/etc", TMP_SYSROOT "/etc", NULL, MS_BIND | MS_REMOUNT | MS_SILENT,
-                 NULL)
-          < 0)
-        err (EXIT_FAILURE, "failed to make writable /etc bind-mount at /sysroot.tmp/etc");
+      if (etc_transient)
+        {
+          /* Do we have a persistent overlayfs for /usr?  If so, mount it now. */
+          g_autofree char *etc_ovldir
+              = g_build_filename (OTCORE_RUN_OSTREE_PRIVATE, "etc-transient", NULL);
+          if (mkdirat (AT_FDCWD, etc_ovldir, 0700) < 0)
+            err (EXIT_FAILURE, "Failed to create %s", etc_ovldir);
+          g_autofree char *upper = g_build_filename (etc_ovldir, "upper", NULL);
+          g_autofree char *work = g_build_filename (etc_ovldir, "work", NULL);
+
+          const char *etc_lower = NULL;
+          if (!find_etc (&etc_lower, &error))
```

Comment:
Isn't this looking in the wrong place (root) for the etc dir? Shouldn't it look in TMP_SYSROOT?

Reply:
Actually, not in sysroot i guess, but deploy dir?

Reply:
Ah, which is CWD, so this is ok.

Reply:
as you already said the deploy dir is the cwd so this should be correct.